experiments/srupp_experiments/data_utils.py

```diff
@@ -421,22 +421,20 @@ def __init__(self, paths, vocab, bsz, bptt, n_nodes=1, rank=0,
 
     def get_sent_stream(self, path):
         sents = self.vocab.encode_file(path, add_double_eos=True)
-        sents = [s for i, s in enumerate(sents) if i % self.n_nodes == self.rank]
         if self.shuffle:
             np.random.shuffle(sents)
         sent_stream = iter(sents)
-
         return sent_stream
 
     def __iter__(self):
         '''
         Iterate over all splits of data *repeatively*
         '''
-        if self.shuffle:
-            np.random.shuffle(self.paths)
-
+        paths = self.paths * self.n_nodes
```

Comment:
why are we multiplying paths by the number of nodes? Does this mean that we intend that each node will load all files during the epoch? Thinking this through for 2 files, f0 and f1, and 2 nodes, does this mean that `paths` would contain `[f0, f1, f0, f1]`? And then when we partition, for node 0, paths_this_node would become `[f0, f0]`?

Reply:
this is to handle that the number of files is not a multiply of the number of GPUs.

Let’s say there 3 files, and 2 GPUs. if we distribute 3 files each time it will be:
GPU 0 gets files: [0, 2]
GPU 1 gets files: [1]
and this repeats.

we make a duplicated file list [0,1,2,0,1,2] so when distributing:
GPU 0 gets files: [0, 2, 1]
GPU 1 gets files: [1, 0, 2]
and this repeats.

what we wanna avoid is:
GPU 0: [0, 2] [0, 2] [0, 2] …
GPU 1: [1][1] [1] [1] ..
this would make examples in file 1 having more weights / appearing more frequently.

your example is ok i believe:
GPU 0: [0, 0] [0, 0] …
GPU 1: [1, 1] [1, 1] …
the GPUs will exchange the training signals / gradients.